### ISSUE TITLE
Enable sklearn tree interface

### ISSUE THREAD

**Issue by crflynn** [grf-labs/grf#991]
In working towards improving `skgrf`, I've added estimator classes for individual trees. I'd like to make the tree classes more compatible with existing analysis tooling for decision trees and random forests like tree structure plotting and shap, for instance.

In order to achieve this compatibility, I think the best way is to fully implement the sklearn Tree interface. Some of this interface can be implemented by deriving from the output result of grf. There are some other attributes that I would like to request here.

Sklearn decision trees implement [a low level API](https://scikit-learn.org/stable/auto_examples/tree/plot_unveil_tree_structure.html#tree-structure) that exposes tree structure and some other descriptive attributes. There are some attributes, like `children_left`, `children_right`, `feature`, and `threshold` which can be derived from the serialized forests' `child_nodes`, `child_nodes`, `split_vars`, and `split_values`, respectively. There are some others like `n_node_samples`, `weighted_n_node_samples`, `value`, and `impurity`, which might also be useful for analysis that I don't believe can currently be derived from serialized forest output. The full set of tree attributes can be found in [this docstring](https://github.com/scikit-learn/scikit-learn/blob/674e7dff128700de7b1e9588b18a78ad6f38f12a/sklearn/tree/_tree.pyx#L494).

I've done a partial implementation so far [here](https://github.com/crflynn/skgrf/pull/50), but I think the rest of the attributes would need to be implemented in C++. I'd be happy to help on implementation if there is mutual interest.

**Comment by erikcs** [comment]
Hi @crflynn, thanks a lot for starting such an excellent Python package. Some additional attributes you mention can be reconstructed from the forest C++ output:

If `n_node_samples` is defined as: the number of training samples used in constructing this tree that reach node i, then [get_tree](https://grf-labs.github.io/grf/reference/get_tree.html) gives an example of how to construct this with traversal: it is the length of the`samples` attribute of a terminal node (note that with default honest splitting, this amounts to the second repopulation subsample).

If `weighted_n_node_samples` is defined as: the sum of sample weights from the previous definition, then this too can be constructed as \sum_{j \in node i} sample_weights_j.

`impurity` is going to be annoying but you can still construct it from the forest output and peripheral information (nuisance components like propensity scores). For a causal forest (or GRF in general) the regression split impurity is calculated from pseudo-outcomes constructed by the `RelabelingStrategy`, for causal forest this amounts to the gradient of tau in the regression `Y = c + tau W` where Y and W are centered outcomes and propensities from the parent node.  (Considering you mention Shap [crossref #986] values as an intended use-case, `impurity` might not be worth prioritizing atm as they play no role in that)

`value` requires more care: recall that GRF takes a different perspective from traditional random forest and does not average predictions over each tree, but instead uses the arising neighborhood from a query sample as a kernel weight (eqn 3 in https://arxiv.org/pdf/1610.01271.pdf) in solving an appropriate estimating equation (section 2.1 in https://arxiv.org/pdf/1610.01271.pdf). The Tree ensemble explainer in [SHAP](https://shap.readthedocs.io/en/latest/generated/shap.explainers.Tree.html#) / [paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7326367/) is built on the first definition.

In the case of a standard regression forest, these two definitions coincide, and you can reconstruct the value for a tree as the mean Y of the node samples. For a causal forest, however, they do not, and we need to decide on an approximation z, where we have that sum_{i}^{B} z_i ~= f, where f is the actual prediction for target sample x and B is the number of trees. I am wondering if we can do a gradient of (19) in https://arxiv.org/pdf/1610.01271.pdf wrt. alpha(x)_i. What are your thoughts @swager / @davidahirshberg / @halflearned? A quick search reveals that the [Causal Forest](https://econml.azurewebsites.net/_autosummary/econml.dml.CausalForestDML.html?highlight=causal%20forest#econml.dml.CausalForestDML.shap_values) in EconML supports Shap values, however it is not clear how the approximation over trees is done, maybe it is worth tagging @vsyrgkanis here if he could share his experience?

**Comment by vsyrgkanis** [comment]
Yeap we simply Shap explain the average of tree predictions for now, even though our predictions are not the average of trees.

for not super deep trees and especially when constraints on the variance of within leaf treatment is imposed this works quite well to pick up the learned patterns. Things get weird as expected when leafs are super small (e.g. 2) and no constraint on the variance of treatment is kmposed

**Comment by vsyrgkanis** [comment]
See microsoft/EconML#464 for exactly a case where this was the issue and the discussion therein.

**Comment by erikcs** [comment]
Thank you @vsyrgkanis. @crflynn, in that case, computing values is still (as a first working version at least) doable w/o changes to the C++ core. For a causal forest, if you let `samples` be a vector of sample ids for a leaf node, then you can get that node's prediction as:

```R
Y.centered = Y[samples] - forest$Y.hat[samples]
W.centered = W[samples] - forest$W.hat[samples]
value = cov(Y.centered, W.centered) / var(W.centered)
```
(where some provisions has to be made for when var(W.centered) = 0, or an empty node, which can happen if `honesty.prune.leaves` is false (it is true by default)). Another approximation could maybe use the implicit function theorem, but since that involves inverting many potentially very large Jacobian matrix, it is perhaps not feasible (and I guess the reason @vsyrgkanis took the first approach).

Since your Python wrapper holds pointers directly to the forest object, all of the operations involved here, like tree traversal, should be practically as fast as native speed since you can call directly into `Tree::find_leaf_nodes(data, samples)` with a Cython wrapper. Feel free to ping for anything.

**Comment by erikcs** [comment]
And let us know @crflynn if you have things you think we should send a PR for, or a PR you would like review input from us on
